Added src/docker/Dockerfile

amihaiemil opened the pull request:
PR regarding issue #82 . As explained by @original-brownbear , first step is to add this file, then build and push the image to Dockerhub, and then specify that image in .rultor.yml .

amihaiemil commented:
@original-brownbear Who has to build the image and push it to dockerhub? I suppose Yegor, since rultor's image is also under his name.

original-brownbear commented:
@amihaiemil yes @yegor256 needs to setup an automated build for this.

dmarkov commented:
@amihaiemil Many thanks, I will find someone to review it before we merge

dmarkov commented:
@mkordas could you please review this pull request

amihaiemil commented:
@mkordas Yegor has to setup automated build of the image on Dockerhub and then I have to specify the image in ``.rultor.yml``. cc: @yegor256

mkordas commented:
@amihaiemil I'm on it

mkordas commented on src/docker/Dockerfile:
```diff
@@ -0,0 +1,129 @@
+#RUN /bin/bash -l -c "gem update"
+#RUN /bin/bash -l -c "gem install nokogiri"
+#RUN /bin/bash -l -c "gem install bundler"
+#RUN echo 'source /usr/share/rvm/scripts/rvm' >> /etc/bash.bashrc
```
@amihaiemil If it's commented out, why not delete it?

amihaiemil replied:
@mkordas I removed commented block.

mkordas commented on src/docker/Dockerfile:
```diff
@@ -0,0 +1,129 @@
+#RUN /bin/bash -l -c "gem install bundler"
+#RUN echo 'source /usr/share/rvm/scripts/rvm' >> /etc/bash.bashrc
+
+# PHP
```
@amihaiemil why do we need PHP here?

amihaiemil replied:
@mkordas @mkordas I removed PHP.

mkordas commented on src/docker/Dockerfile:
```diff
@@ -0,0 +1,129 @@
+#   echo "zend_extension=xdebug.so" > /etc/php5/cli/conf.d/xdebug.ini
+
+# Java
+RUN apt-get install -y default-jdk
```
@amihaiemil what's the Java version here?

amihaiemil replied:
@mkordas According to this http://packages.ubuntu.com/trusty/default-jdk , the default jdk of ubuntu 14.04 is 1.7-51 , but personally I have no guarantee. The author said to reuse Dockerfile from rultor tag 1.61.3, so that's what I did. 

Do you think another specific jdk should be used?

mkordas replied:
@original-brownbear what do you think, should we lock Java version here, or `default-jdk` is fine?

original-brownbear replied:
@mkordas lock for sure! there is simply no upside to not locking in the build versions 100%. Everything but doing so adds needless risk!

mkordas replied:
@original-brownbear thanks for confirmation

mkordas commented on src/docker/Dockerfile:
```diff
@@ -0,0 +1,129 @@
+  ln -s /usr/share/maven3/bin/mvn /usr/bin/mvn && \
+  rm -rf maven3_3.2.1-0~ppa1_all.deb
+
+# LaTeX
```
@amihaiemil I don't think we need LaTeX here

amihaiemil replied:
@mkordas I removed LaTeX.

mkordas commented on src/docker/Dockerfile:
```diff
@@ -0,0 +1,129 @@
+
+# Basic Linux tools
+RUN apt-get update
+RUN apt-get install -y wget bcrypt curl
```
@amihaiemil please see https://docs.docker.com/engine/userguide/eng-image/dockerfile_best-practices/ - there is better way to call `apt-get`

amihaiemil replied:
@mkordas done

mkordas commented on src/docker/Dockerfile:
```diff
@@ -0,0 +1,129 @@
+RUN sudo npm install -g npm
+RUN ln -s /usr/bin/nodejs /usr/bin/node
+
+# One more time, right before the end
```
@amihaiemil why this is needed?

amihaiemil replied:
@mkordas from what I see in ``man apt-get`` those lines update the used sources. Here, I suppose they are used as a safety net, to check for updates one last time. If you think they can be safely deleted, I have nothing against it, but I am not sure about it. Since I'm no expert in linux, I simply did as the issue author said (reused the dockerfile of rultor, without the pull and build of rultor).

mkordas replied:
@amihaiemil OK, let's ask the author

mkordas replied:
@original-brownbear do you think we should have these lines? what's the purpose of them?

original-brownbear replied:
@mkordas No they are a *big* anti-pattern when using Docker. You should never run any `apt-get upgrade` in a container build. This simply causes you to become reliant on the apt package maintainer to stay backwards compatible. Also `apt-get upgrade` can actually break the build one some base images (short explanation: it may require restarting a service, this is disabled in Debian based base images => broken ).
This is already removed from the Rultor image itself too :)

=> just don't do it lock all versions :)

mkordas replied:
@original-brownbear I've learnt something new, thank you

mkordas commented:
@amihaiemil please see inital review. I think we should begin with minimal image instead.

mkordas commented:
@amihaiemil I'm back

mkordas commented:
@original-brownbear I need hour advice here - should we clean the things up here or leave as it is? E.g. nodejs is clearly not needed, but @amihaiemil did what he was told to do - just copy the image.

mkordas commented on src/docker/Dockerfile:
```diff
@@ -0,0 +1,97 @@
+# Copyright (c) 2012-2015, jcabi.com
```
@amihaiemil now we need to upload this image to Docker Hub

mkordas replied:
@amihaiemil and of course update `.rultor.yml` to use it

original-brownbear commented:
@mkordas sure :) Damn looking at that ticket again it's worded in a little unfortunate manner :)

But agreed, simply copying the image was my suggested quick fix, obviously it's not even close to being an optimal solution. In a sense it is my mistake not that of the performer that the description leads to a rather iffy solution now.
Maybe ask the architect here too, but I think it would be "fair" in the sense that the performer only has to resolve the actual stated issue, not an issue refined by the author in hindsight to simply:
* Remove the `apt-get update && apt-get upgrade` stuff because it's really bad
* Add a puzzle asking to clean up the image removing unnecessary things from it and locking in all versions 100%.

Makes sense ? :)

amihaiemil commented:
@mkordas As the author suggested, I removed the apt-get upgrate/update commands and added puzzle for packages version lock. Yegor has to set automated build of the image on Dockerhub, not me. After he is done, I willl specify the image in ``.rultor.yml``.

mkordas commented:
@amihaiemil I will check it now

mkordas commented on src/main/java/com/jcabi/aether/Aether.java:
```diff
@@ -95,6 +95,9 @@
  * @todo #69:30min Fix UnsupportedClassVersionError in CL build for Open JDK6.
+ * @todo #82:30min Clean unnecessary docker setup (src/docler/Dockerfile) and set
+ *  versions to the used packages (e.g. lock java version instead of default
+ *  jdk). Discuss with architect.
```
@amihaiemil I don't think that much discussion is needed here. I would stress that the goal here is to have minimal docker image with all versions locked that is enough to build this project.

amihaiemil replied:
@mkordas I meant discuss the needed versions. Because for sure, the first questin will be "OK, but what versions exactly are needed?"

mkordas commented:
@amihaiemil so we need additional puzzle just for Yegor, right? How we are going to track these tasks?

mkordas commented:
@amihaiemil please see my comments above

amihaiemil commented:
@mkordas I detailed the todo... about the Dockerhub thing, I was hoping Yegor can do it directly after you approve the Dockerfile. Is it a lot of work, or the Dockerfile has to be on the master branch?

I think if we create a todo for that then and then anotherone for me to update ``.rultor.yml``, this task will take way too long + like you said, it needs particular tracking and besides, there is already another issue waiting for this to be completed.

amihaiemil commented:
@mkordas I say we merge this, ask Yegor to set the automated build in Dockerhub and then I make a 2nd PR with the ``.rultor.yml`` change. What do you say?

mkordas commented:
@amihaiemil I'm back again

mkordas commented on src/main/java/com/jcabi/aether/Aether.java:
```diff
@@ -95,6 +95,10 @@
+ * @todo #82:30min Clean unnecessary docker setup (src/docler/Dockerfile),
+ *  make the configuration minimal. Set versions to the used packages
+ *  (e.g. lock java version instead of default jdk). Discuss with architect
+ *  the needed versions.
```
@amihaiemil are you sure we need that discussion? We just need latest Java 7, latest Maven and latest versions of maybe 2-3 more tools, as this project is very simple.

mkordas commented on src/main/java/com/jcabi/aether/Aether.java:
```diff
@@ -95,6 +95,10 @@
  * @todo #69:30min Fix UnsupportedClassVersionError in CL build for Open JDK6.
+ * @todo #82:30min Clean unnecessary docker setup (src/docler/Dockerfile),
+ *  make the configuration minimal. Set versions to the used packages
+ *  (e.g. lock java version instead of default jdk). Discuss with architect
```
@amihaiemil please also add an info that after changes it needs to be uploaded to Docker Hub and built there

mkordas commented:
@amihaiemil yeah, Dockerfile don't need to me on master, so let's try. But Yegor travels a lot and he may be unavailable for a few days.

mkordas commented:
@amihaiemil I'm ok to do 2 PRs for this issue, second one after image will be available

mkordas commented:
@amihaiemil see my 2 comments about puzzle, except that all good to go

amihaiemil commented:
@mkordas Done. If the Dockerfile does not need to be on the master branch, let's wait for Yegor to set it up on DockerHub and then I'll modify ``.rultor.yml`` and we can merge and close this... 
We could also make 2 PRs, but we must wait for Yegor's approval to merge this anyway, so it's the same thing,

amihaiemil commented:
@yegor256 Can you please set up the automated built of ``src/docker/Dockerfile`` on DockeHub? Then I can specify the image in ``.rultor.yml`` and we can merge this. Thanks.

mkordas commented:
@yegor256 we need your help here

amihaiemil commented:
@yegor256 ping

yegor256 commented:
@amihaiemil done: https://hub.docker.com/r/yegor256/jcabi-aether/

mkordas commented:
@yegor256 thanks
@amihaiemil please proceed

amihaiemil commented:
@mkordas done

mkordas commented:
@amihaiemil looks ok!

mkordas commented:
@rultor merge

rultor commented:
> @rultor merge

@mkordas Thanks for your request. @yegor256 Please confirm this.

mkordas commented:
@yegor256 we need your merge here

yegor256 commented:
@rultor try to merge

rultor commented:
> @rultor try to merge

@yegor256 OK, I'll try to merge now. You can check the progress of the merge [here](http://www.rultor.com/t/8026-209113293)

rultor commented:
> @rultor try to merge

@amihaiemil @yegor256 Oops, I failed. You can see the full log [here](http://www.rultor.com/t/8026-209113293) (spent 3min)

```
"

Tests in error: 
  AetherTest.findsAndLoadsArtifacts:117 » DependencyResolution failed to load 'c...

Tests run: 16, Failures: 1, Errors: 1, Skipped: 0

[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 28.825 s
[INFO] Finished at: 2016-04-12T21:40:33+00:00
[INFO] Final Memory: 32M/368M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:2.17:test (default-test) on project jcabi-aether: There are test failures.
[ERROR] 
[ERROR] Please refer to /home/r/repo/target/surefire-reports for the individual test results.
[ERROR] -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:2.17:test (default-test) on project jcabi-aether: There are test failures.

Please refer to /home/r/repo/target/surefire-reports for the individual test results.
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:212)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:153)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:145)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:116)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:80)
	at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build(SingleThreadedBuilder.java:51)
	at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:128)
	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:307)
	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:193)
	at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:106)
	at org.apache.maven.cli.MavenCli.execute(MavenCli.java:863)
	at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:288)
	at org.apache.maven.cli.MavenCli.main(MavenCli.java:199)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:289)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:229)
	at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:415)
	at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:356)
Caused by: org.apache.maven.plugin.MojoFailureException: There are test failures.

Please refer to /home/r/repo/target/surefire-reports for the individual test results.
	at org.apache.maven.plugin.surefire.SurefireHelper.reportExecution(SurefireHelper.java:82)
	at org.apache.maven.plugin.surefire.SurefirePlugin.handleSummary(SurefirePlugin.java:195)
	at org.apache.maven.plugin.surefire.AbstractSurefireMojo.executeAfterPreconditionsChecked(AbstractSurefireMojo.java:861)
	at org.apache.maven.plugin.surefire.AbstractSurefireMojo.execute(AbstractSurefireMojo.java:729)
	at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:134)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:207)
	... 20 more
[ERROR] 
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
container d4aede0b35e3347c3ca2b6b2dd4a3b62d2b24634523e676a28afc910dc8fcf2b is dead
Tue Apr 12 21:40:37 UTC 2016

```

mkordas commented:
@amihaiemil the merge failed with
```
java.lang.IllegalArgumentException: failed to load 'com.jcabi:jcabi-x:jar:0.0.0 (test)' from ["Maven (http://repo.maven.apache.org/maven2, releases+snapshots) without authentication", "Maven (http://repo.maven.apache.org/maven2, releases+snapshots) without authentication", "Maven (http://repo.maven.apache.org/maven2, releases+snapshots) without authentication", "Maven (http://repo.maven.apache.org/maven2, releases+snapshots) without authentication"] into /tmp/junit4378744301392343738/junit6907336093823982690
```
I'm not sure about the reason though...

original-brownbear commented:
@mkordas I think it fails for the very same reason it failed to begin with.
The newest Rultor image doesn't work with this project, now the `.rultor.yml` isn't yet updated in `master` so that it uses the new image => still fails :/

mkordas commented:
@original-brownbear so what can we do now? merge by a button?

original-brownbear commented:
@mkordas yes, Yegor needs to do this.

mkordas commented:
@yegor256 please merge with a button as explained by @original-brownbear above

yegor256 commented:
@amihaiemil merged

amihaiemil commented:
@yegor256 Thanks

mkordas commented:
@original-brownbear @yegor256 maybe let's try to do the deployment with a new image

mkordas commented:
@rultor deploy

rultor commented:
> @rultor deploy

@mkordas Thanks for your request. @yegor256 Please confirm this.

original-brownbear commented:
@mkordas yea that should work, otherwise we need to reopen the issue I'm afraid.

dmarkov commented:
@mkordas Thank you, I have added **52 mins** to your account in payment/transaction "`570f9b99c1b2190a20000319`", time consumed: 194 hours and 57 mins; you're getting extra minutes here (c=37); +52 added to your rating, current score is: [+8571](http://www.netbout.com/b/36080?open=rating)

dmarkov commented:
@rultor deploy now pls

rultor commented:
> @rultor deploy now pls

@dmarkov OK, I'll try to deploy now. You can check the progress [here](http://www.rultor.com/t/8026-209941932)

rultor commented:
> @rultor deploy now pls

@amihaiemil @dmarkov Oops, I failed. You can see the full log [here](http://www.rultor.com/t/8026-209941932) (spent 24s)

```
+ set -o pipefail
++ dirname ./run.sh
+ cd .
+ echo 359
+ echo '1.61.8 b552866'
1.61.8 b552866
+ date
Thu Apr 14 13:33:07 UTC 2016
+ uptime
 13:33:07 up 13 days, 18:24,  0 users,  load average: 4.58, 5.31, 4.13
+ head=git@github.com:jcabi/jcabi-aether.git
+ head_branch=master
+ vars=('--env=MAVEN_OPTS=-XX:MaxPermSize=256m -Xmx1g' '--env=head=git@github.com:jcabi/jcabi-aether.git' '--env=head_branch=master')
+ image=yegor256/jcabi-aether
+ directory=
+ scripts=('sudo gem install --no-ri --no-rdoc pdd' ';' 'sudo gem install --no-ri --no-rdoc est' ';' 'mvn clean deploy -Pqulice --errors --settings ../settings.xml' ';' 'mvn clean' ';' 'pdd --source=$(pwd) --verbose --file=jcabi-aether.xml' ';' 's3cmd --no-progress put jcabi-aether.xml --config=../s3cfg s3://pdd.teamed.io/jcabi-aether.xml' ';' 'est --dir=./est --verbose --file=jcabi-aether.xml --format=xml' ';' 's3cmd --no-progress put jcabi-aether.xml --config=../s3cfg s3://est.teamed.io/jcabi-aether.xml' ';')
+ squash=false
+ ff=default
+ rebase=false
+ container=jcabi_jcabi-aether_83
+ as_root=false
+ mkdir -p /home/rultor/.ssh
+ echo -e 'Host github.com\n\tStrictHostKeyChecking no\n'
+ chmod 600 /home/rultor/.ssh/config
+ git clone git@github.com:jcabi/jcabi-aether.git repo
Cloning into 'repo'...
+ cd repo
+ git config user.email me@rultor.com
+ git config user.name rultor
+ '[' -z 'sudo gem install --no-ri --no-rdoc pdd' ']'
+ cd ..
+ cat
+ '[' false = true ']'
+ cat
+ chmod a+x entry.sh
+ cat
+ echo 'sudo gem install --no-ri --no-rdoc pdd' ';' 'sudo gem install --no-ri --no-rdoc est' ';' 'mvn clean deploy -Pqulice --errors --settings ../settings.xml' ';' 'mvn clean' ';' 'pdd --source=$(pwd) --verbose --file=jcabi-aether.xml' ';' 's3cmd --no-progress put jcabi-aether.xml --config=../s3cfg s3://pdd.teamed.io/jcabi-aether.xml' ';' 'est --dir=./est --verbose --file=jcabi-aether.xml --format=xml' ';' 's3cmd --no-progress put jcabi-aether.xml --config=../s3cfg s3://est.teamed.io/jcabi-aether.xml' ';'
+ rm -rf .gpg
+ cd repo
+ docker_when_possible
+ true
++ sed 's/ /\n/g'
++ tail -n 1
++ uptime
+ load=4.16
++ bc
++ echo 4.16 '>' 30
+ '[' 0 -eq 1 ']'
+ echo 'load average is 4.16, low enough to run a new Docker container'
load average is 4.16, low enough to run a new Docker container
+ break
+ cd ..
+ '[' -n '' ']'
+ use_image=yegor256/jcabi-aether
+ docker pull yegor256/jcabi-aether
Using default tag: latest
Pulling repository docker.io/yegor256/jcabi-aether
Tag latest not found in repository docker.io/yegor256/jcabi-aether
'cid' file is absent, container wasn't started correctly

```

amihaiemil commented:
@original-brownbear Do you happen to know what went wrong? It seems to me some config on Yegor's DockerHub? He needs to define the 'latest' tag?

original-brownbear commented:
@amihaiemil yes exactly, `latest` is missing.

amihaiemil commented:
@yegor256 Can you pls define the ``latest`` tag, so the build passes? Issue #80  depends on this since the PR for it cannot be merged. Thanks.

yegor256 commented:
@amihaiemil it is "defined", as far as I understand:

<img width="1278" alt="screen shot 2016-04-17 at 7 21 18 pm" src="https://cloud.githubusercontent.com/assets/526301/14588762/a4a3c0f2-04d1-11e6-8302-53b1f36fa2c8.png">

If you need me to do something else, please give more detailed instructions

original-brownbear commented:
@yegor256 The build for this thing failed ...
<img width="1079" alt="screen shot 2016-04-17 at 19 24 09" src="https://cloud.githubusercontent.com/assets/6490959/14588784/4f327748-04d2-11e6-8a83-4deee1ea2b24.png">

maybe hit the trigger button there and see if it works on retry before the Dockerfile issue gets reopened ... ( maybe we get lucky and it was just a random apt issue, seems it didn't find some package from the Dockerhub logs).

yegor256 commented:
@original-brownbear triggered it a second ago... let's wait

original-brownbear commented:
@yegor256 it cannot work ... The `Dockerfile` contains this :
```
RUN apt-get update & apt-get install -y \
```

one `&` where we need `&&` ... don't think this warrants a ticket really right?, maybe you just push the fix for that?

yegor256 commented:
@original-brownbear fixed in 5f445b0

original-brownbear commented:
@yegor256 phew ... you fixed it in a different instance 2 left still:
https://github.com/jcabi/jcabi-aether/blob/master/src/docker/Dockerfile#L57
https://github.com/jcabi/jcabi-aether/blob/master/src/docker/Dockerfile#L44

how on earth did we all miss that when reviewing this ...

yegor256 commented:
@original-brownbear fixed in 9b81a37

original-brownbear commented:
@yegor256 thanks! looks good now, build on Dockerhub worked, I guess we can try merging again :)

yegor256 commented:
@original-brownbear cool!
